pkg/compose/run.go

```diff
@@ -116,8 +116,13 @@ func applyRunOptions(project *types.Project, service *types.ServiceConfig, opts
 	if len(opts.Environment) > 0 {
 		env := types.NewMappingWithEquals(opts.Environment)
 		projectEnv := env.Resolve(func(s string) (string, bool) {
-			if _, ok := service.Environment[s]; ok {
-				return "", false
+			cmdLineVar, ok := env[s]
+			if ok && cmdLineVar == nil {
+				v, ok := project.Environment[s]
+				return v, ok
+			}
+			if sEnv, ok := service.Environment[s]; ok && sEnv != nil {
+				return *sEnv, ok
```

Comment:
I'm having a hard time following this.

I _THINK_ it should be simplified/shortened to:
```go
v, ok := project.Environment[s]
return v, ok
```

My understanding is that `env` is the `cmdEnv`, and `Resolve` will call your `lookupFn` for any values that have `nil`, right?

So the immediate lookup from from `env` (aka `cmdEnv`) is guaranteed to be `nil`.
As a result, we'll always fall back to returning whatever is in the project (OS/host) env and never attempt to read from the service env here...which I think is what we want!
User did `compose run --env FOO` which explicitly means "read this value from my OS env".

If I'm missing some nuance here, some comments about the why/how you end up in each of these cases and how that relates to precedence would be super helpful.

TL;DR I think this code is _technically_ correct, but very roundabout?

Reply:
Ahhh!!! Nice catch!!!
That was an old "bad good idea" of when using `-e UNSETVAR` and it couldn't be found in `project.Environment` it could check in the `service:environment` section of the Compose file.
But... As you said, that's bad! Specially when the user imperatively used the `-e` option to make its usage explicit.